src/H5.c

```diff
@@ -208,15 +208,6 @@ H5_init_library(void)
      */
     if (!H5_dont_atexit_g) {
 
```

Comment:
Why is this change necessary?

Reply:
In order to be able to clean up error records properly, H5TS needs to perform the cleanup before `H5E_term_package()`. It would be possible to split `H5TS_term_package()` into a part that runs at process exit here and a part that runs during library shutdown so it can precede H5E, but I figured it was better practice to keep it all in one place.

Reply:
What part of the H5TS code needs cleaned up before the H5E shuts down?

Reply:
It's not the H5TS code that needs to be cleaned up before H5E per se; it's part of H5E that needs to be cleaned up before  H5E shuts down, and only H5TS has the information necessary to perform the clean up.

As the library is presently, multiple pieces of memory don't get cleaned up properly during the execution of `ttsafe_error_stacks`:

- the error descriptions allocated during `H5E__set_stack_entry()`
- the temporary function and files strings allocated during `H5Epush2()`

If you just add `H5E__clear_stack()` calls to the thread info cleanup in `H5TS_term_package()`, then the H5I_dec_ref() calls fails due to having already cleaned up the error classes/messages. (This then cascades into an assertion failure, since it tries to push an error after having already cleaned up H5E.)

Reply:
Thanks, that helps.   I see what you are thinking about now.

This is a more fundamental problem, which we're going to have to partially solve with "don't do that" instructions to application developers.

For instance, it's may not be supported for one thread to shutdown the library while other threads exist, since they may be using or about to use the library.  And since we can't detect when an application creates or joins threads, it's likely that we will have to take a solution like MPI: you must initialize the library before your application creates threads and can only shut it down (possibly via an atexit() callback) after all the threads are joined.

Otherwise, I don't think there's a way to correctly get the 'once' initialization to work correctly.

Reply:
All that said, I think that the H5E__clear_stack() call _does_ need to be made within the H5TS__tinfo_destroy() routine, so that the error stack can be cleaned up when a thread is joined.

Reply:
I don't think it belongs in the other places, and I don't think the split shutdown that you are describing for releasing the main thread's error stack is necessary, since the main thread will call H5E_term_package() to release it.

Hmm - might have to refcount (with an atomic int) the number of threads active in the library, in order to know when the last one (the main one) is shutting down and the error stack destroy should be skipped.

Reply:
Probably worthwhile to set up a call to talk, if you'd like.

Reply:
I'm not sure if "don't do that" instructions will suffice here - this problem comes up whenever the last API routine used in a thread raises an HDF5 error.  Even if the applications is nice and closes all of its threads before terminating the library, the memory will still be lost, because all the per-thread cleanup does is add its thread info node to the global free list, which is still subject to the cleanup ordering problems.

I'll reach out to set up a time to talk through this.